base/stream.jl

```diff
@@ -1215,20 +1215,95 @@ i.e. data to be read from [`stdin`](@ref) may be written to `wr`.
 """
 redirect_stdin
 
-for (F,S) in ((:redirect_stdin, :stdin), (:redirect_stdout, :stdout), (:redirect_stderr, :stderr))
-    @eval function $F(f::Function, stream)
-        STDOLD = $S
-        local ret
-        $F(stream)
-        try
-            ret = f()
-        finally
-            $F(STDOLD)
-        end
-        ret
+
+"""
+    redirect(;stdin=stdin, stderr=stderr, stdout=stdout)
+
+Redirect a subset of the streams `stdin`, `stderr`, `stdout`.
+"""
+function redirect(;stdin=nothing, stderr=nothing, stdout=nothing)
+    stdin  === nothing || redirect_stdin(stdin)
+    stderr === nothing || redirect_stderr(stderr)
+    stdout === nothing || redirect_stdout(stdout)
+end
+
+"""
+    redirect(f; stdin=nothing, stderr=nothing, stdout=nothing)
+
+
+Redirect a subset of the streams `stdin`, `stderr`, `stdout`,
+call `f()` and restore each stream.
+
+Possible values for each stream are:
+* `nothing` indicating the stream should not be redirected.
+* `path::AbstractString` redirecting the stream to the file at `path`.
+* `io` an `IOStream`, `TTY`, `Pipe`, socket, or `devnull`.
+
+```julia
+julia> redirect(stdout="stdout.txt", stderr="stderr.txt") do
+           print("hello stdout")
+           print(stderr, "hello stderr")
+       end
+
+julia> read("stdout.txt", String)
+"hello stdout"
+
+julia> read("stderr.txt", String)
+"hello stderr"
+```
+
+!!! compat "Julia 1.6"
+    `redirect` requires Julia 1.6 or later.
+"""
+function redirect(f; stdin=nothing, stderr=nothing, stdout=nothing)
+
+    function resolve(new::Nothing, oldstream, mode)
+        (new=nothing, close=false, old=nothing)
+    end
+    function resolve(path::AbstractString, oldstream,mode)
+        (new=open(path, mode), close=true, old=oldstream)
+    end
+    function resolve(new, oldstream, mode)
+        (new=new, close=false, old=oldstream)
+    end
+
+    same_path(x, y) = false
+    same_path(x::AbstractString, y::AbstractString) = x == y
+    if same_path(stderr, stdin)
+        throw(ArgumentError("stdin and stderr cannot be the same path"))
+    end
+    if same_path(stdout, stdin)
+        throw(ArgumentError("stdin and stdout cannot be the same path"))
+    end
+
+    if same_path(stderr, stdout)
```

Comment:
@JeffBezanson If the user passes the same path for `stdout` and `stderr` the `io` is opened only once and everything works fine. 
If the user passes e.g. two distinct `io` that point to the same file, that is still a problem. But the same is true on master e.g:
```julia
single_path = "single_file.txt"
open(single_path, "w") do new_stdout
     open(single_path, "w" do new_stderr
         redirect_stdout(new_stdout) do
             redirect_stderr(new_stderr) do
                   f()
             end
        end
    end
end
```
Should I document this? Should I try to handle it (I think an `IOStream` remembers its path)?

Reply:
I documented this.